[Merged by Bors] - feat(Data/Finset/Max): `min'_pair`, `max'_pair`

jsm28 opened the pull request:
Add the lemmas:

```lean
lemma min'_pair (a b : α) : min' {a, b} (insert_nonempty _ _) = min a b := by
```

```lean
lemma max'_pair (a b : α) : max' {a, b} (insert_nonempty _ _) = max a b := by
```

and corresponding `min_pair` and `max_pair`. So that the primed lemmas can be proved by `simp`, like the unprimed ones, make `min'_insert` and `max'_insert` into `simp` lemmas, and swap the arguments to `min` and `max` on their RHS, also corresponding more closely to the unprimed lemmas `min_insert` and `max_insert`.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

jsm28 committed:
feat(Data/Finset/Max): `min'_pair`, `max'_pair`

Add the lemmas:

```lean
@[simp] lemma min'_pair (a b : α) : min' {a, b} (insert_nonempty _ _) = min a b := by
```

```lean
@[simp] lemma max'_pair (a b : α) : max' {a, b} (insert_nonempty _ _) = max a b := by
```

eric-wieser commented on Mathlib/Data/Finset/Max.lean:
```diff
@@ -251,6 +251,20 @@ lemma min'_eq_iff (a : α) : s.min' H = a ↔ a ∈ s ∧ ∀ (b : α), b ∈ s
   ⟨(· ▸ ⟨max'_mem _ _, le_max' _⟩), fun h ↦ le_antisymm (max'_le _ _ _ h.2) (le_max' _ _ h.1)⟩
 
+@[simp] lemma min'_pair (a b : α) : min' {a, b} (insert_nonempty _ _) = min a b := by
+  rw [Finset.min'_eq_iff]
+  simp
+  rcases le_or_gt a b with h | h
+  · exact .inl h
+  · exact .inr h.le
```
If you move this down the file the proof is just
```suggestion
  rw [Finset.min'_insert, Finset.min'_singleton, min_comm]
```

eric-wieser replied:
The `min_comm` suggests to me that `min'_insert` might be stated backwards, though I haven't thought much about it.

jsm28 committed:
Update Mathlib/Data/Finset/Max.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented:
Perhaps the right thing to do here is tag `min'_insert` `@[simp]` and add no new lemmas?

jsm28 committed:
Move down and shorten `max'_pair` likewise

jsm28 commented:
Note that since `min'_insert` requires the hypothesis `H`, its use as a `simp` lemma will be more limited, to when `simp` can prove `H` or a suitable hypothesis is passed to `simp`.

jsm28 committed:
fix(Data/Finset/Max): Use `DecidableEq` for `insert` lemmas

These `insert` in these lemma statements uses a `DecidableEq` instance
derived from `LinearOrder`; make them syntactically more general by
passing in a `DecidableEq` instance separately.  (Some `insert` lemmas
in this file already take `DecidableEq` hypotheses.)

jsm28 committed:
Merge branch 'min_insert_decidableeq' into finset_minmax_pair

jsm28 committed:
Make `insert` lemmas `@[simp]`, prove new lemmas with `simp`

jsm28 commented:
It seems less clear what the unprimed lemmas should state (where do the coercions to `WithTop` / `WithBot` go on the RHS?).

jsm28 committed:
Merge branch 'master' into min_insert_decidableeq

jsm28 committed:
Merge branch 'min_insert_decidableeq' into finset_minmax_pair

jsm28 committed:
Merge branch 'master' into min_insert_decidableeq

jsm28 committed:
Merge branch 'min_insert_decidableeq' into finset_minmax_pair

jsm28 committed:
Add unprimed lemmas

jsm28 committed:
Merge branch 'master' into min_insert_decidableeq

jsm28 committed:
Merge branch 'min_insert_decidableeq' into finset_minmax_pair

jsm28 committed:
Remove dependency on #29422

jsm28 committed:
Update Mathlib/Data/Finset/Max.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

jsm28 committed:
`max'_insert`, `min'_insert` fixes
